### PR TITLE
Port to jbuilder

### PR DESCRIPTION
**Pull request by rgrinberg**
Porting to jbuilder has the following consequences:

* We now have separate opam packages for sub libraries (yay!)
* Our builds are much faster
* We don't need to check in generated garbage

One side effect:

the rename from s.mli to s.ml. jbuilder doesn't support mli only modules
and there's no need for them anyway.

To use jbuilder, install it and try the following:

```
$ jbuilder build @install
$ jbuilder build @runtest
```

cc @hcarty

### CODE REVIEW

File: redis-sync.opam [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+bug-reports: "https://github.com/0xffea/ocaml-redis/issues"
+license: "BSD3"
+tags: []
+dev-repo: "https://github.com/0xffea/ocaml-redis.git"
+build: [
+  ["jbuilder" "build" "--only-packages" "redis-sync" "--root" "."
+   "-j" jobs "@install"]
+]
+depends: [
+  "redis-" {build}
```
Comment (hcarty) [inline redis-sync.opam]: `redis-` <- typo?

**Comment by hcarty** [comment]
Apologies for some relatively naive questions:

1. Should all of the `*.opam` files depend on jbuilder directly, since they use it explicitly?
2. Why is the dependency on `redis` from the no-longer-sub-packages a `{build}` dependency?

My only other concern/question is if it's worth waiting a bit longer for jbuilder to settle out.  There have been a lot of changes over the last few jbuilder betas.

**Comment by rgrinberg** [comment]
1. I debated about adding jbuilder explicitly. But my thinking is that a sub opam package will always share the build system of the parent package so it isn't necessary. I guess it's not bad to be explicit and just add it.
2. The `{build}` marker on redis was accidental.

**Comment by rgrinberg** [comment]
> My only other concern/question is if it's worth waiting a bit longer for jbuilder to settle out. There have been a lot of changes over the last few jbuilder betas.

Yeah, I think it's fine to wait for 1.0 before merging and releasing this. Note that most of the flux in jbuilder has been with relation to ppx handling. So packages like this are quite safe.

**Comment by hcarty** [comment]
Thanks for addressing all of my questions so quickly! Waiting for 1.0 is probably prudent. Other than that the structural changes look good to me.

Travis seems unhappy, maybe some config change required there? That can probably come once this is merged/about to be merged though.

**Comment by rgrinberg** [comment]
Looks like travis is broken for unrelated reasons:

> # Error: Findlib package "lwt.unix" not found.

This is likely from that recent split in lwt. We should depend on lwt-unix explicitly.

Also travis should test against 4.04 explicitly now since 4.05 is about to come out.

**Comment by aluuu** [comment]
Hi @rgrinberg,

Thank you for the pull request. Sorry for such delayed answer. 

I'll look closer into CI error.

**Pull request merged**